pkg/operator/boundsatokensignercontroller/controller.go

```diff
@@ -0,0 +1,303 @@
+	// added to the operator's signing key secret and set to now + the
+	// ready interval. Only once that interval has elapsed will the
+	// contents of the operator secret be promoted to the operand
+	// secret for use by the apiserver. This interval is intended to
+	// ensure that apiservers have a chance to update to the latest
+	// public key configmap that can validate tokens issued by the new
+	// private key. If the secret was promoted immediately, and new
+	// tokens were issued by its private key, apiserver instances that
+	// did not have the latest public key would not be able to
+	// validate those new tokens.
+	TokenReadyAnnotation = "kube-apiserver.openshift.io/ready-to-use"
+	readyInterval        = 5 * time.Minute
+
+	CertConfigMapName = "bound-sa-token-signing-certs"
+)
+
+// BoundSATokenSignerController manages the keypair used to sign bound
+// tokens and the key bundle used to verify them.
+//
+// This code is derived from the controller used by the
+// kube-controller-manager-operator to manage the key material for the
+// legacy sa token signer.
+type BoundSATokenSignerController struct {
+	secretClient    corev1client.SecretsGetter
+	configMapClient corev1client.ConfigMapsGetter
+	eventRecorder   events.Recorder
+
+	cachesSynced []cache.InformerSynced
+
+	// queue only ever has one item, but it has nice error handling backoff/retry semantics
+	queue workqueue.RateLimitingInterface
+}
+
+func NewBoundSATokenSignerController(
+	kubeInformersForNamespaces v1helpers.KubeInformersForNamespaces,
+	kubeClient kubernetes.Interface,
+	eventRecorder events.Recorder,
+) *BoundSATokenSignerController {
+
+	ret := &BoundSATokenSignerController{
+		secretClient:    v1helpers.CachedSecretGetter(kubeClient.CoreV1(), kubeInformersForNamespaces),
+		configMapClient: v1helpers.CachedConfigMapGetter(kubeClient.CoreV1(), kubeInformersForNamespaces),
+		eventRecorder:   eventRecorder.WithComponentSuffix("bound-sa-token-signer-controller"),
+
+		cachesSynced: []cache.InformerSynced{
+			kubeInformersForNamespaces.InformersFor(operatorNamespace).Core().V1().Secrets().Informer().HasSynced,
+			kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().Secrets().Informer().HasSynced,
+			kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().ConfigMaps().Informer().HasSynced,
+		},
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "BoundSATokenSignerController"),
+	}
+
+	kubeInformersForNamespaces.InformersFor(operatorNamespace).Core().V1().Secrets().Informer().AddEventHandler(ret.eventHandler())
+	kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().Secrets().Informer().AddEventHandler(ret.eventHandler())
+	kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().ConfigMaps().Informer().AddEventHandler(ret.eventHandler())
+
+	return ret
+}
+
+func (c *BoundSATokenSignerController) sync() error {
```

Comment:
please break this function into logical bits with no data based between them.  I see 

1. create a secret if needed. depends on nothing outside the secrets
2. update configmap if needed.  This can retrieve the current secret.  A stale lister will always get another event notification, so staleness of a cache doesn't matter and this can then be contained.
3. promoting a key.  This is based on the current secret in the operator namespace, the current secret in the operand namespace, the revisions on the nodes, the content of the configmaps on those nodes.  All of which can be cached.

If any step has an error, the other steps should still be run.

Reply:
Done